flowmachine/flowmachine/features/subscriber/contact_modal_location_distance.py

```diff
@@ -0,0 +1,189 @@
+    """
+    This class calculates statistics for the distance between subscriber's own modal
+    location and its contacts' modal location.
+
+    Parameters
+    ----------
+    start, stop : str
+         iso-format start and stop datetimes
+    statistic : {'count', 'sum', 'avg', 'max', 'min', 'median', 'mode', 'stddev', 'variance'}, default 'sum'
+        Defaults to sum, aggregation statistic over the durations.
+    method : str, default 'last'
+        The method by which to calculate the location of the subscriber.
+        This can be either 'most-common' or last. 'most-common' is
+        simply the modal location of the subscribers, whereas 'lsat' is
+        the location of the subscriber at the time of the final call in
+        the data.
+    contact_balance: features.subscriber.ContactBalance, default None.
+        An instance of ContactBalance. If an instance is not provide, the
+        ContactBalance is instantiated with the same parameters as for the
+        features.subscriber.ModalLocation
+    hours : 2-tuple of floats, default 'all'
+        Restrict the analysis to only a certain set
+        of hours within each day.
+    subscriber_subset : str, list, flowmachine.core.Query, flowmachine.core.Table, default None
+        If provided, string or list of string which are msisdn or imeis to limit
+        results to; or, a query or table which has a column with a name matching
+        subscriber_identifier (typically, msisdn), to limit results to.
+    direction : {'in', 'out', 'both'}, default 'out'
+        Whether to consider calls made, received, or both. Defaults to 'out'.
+    column_name : str
+        Optionally specify a non-default column name. Required if level is 'polygon'.
+
+    Example
+    -------
+
+    >>> s = ContactModalLocationDistance("2016-01-01", "2016-01-03", statistic="avg")
+    >>> s.get_dataframe()
+
+        subscriber           value
+    gwAynWXp4eWvxGP7    298.721500
+    GnyZMedmKQ4X78Wa    290.397556
+    BKMy1nYEZpnoEA7G     78.919136
+    m4L326vrwE6elJxQ    249.033988
+    NG1km5NzBg5JD8nj    188.679378
+               ...           ...
+    """
+
+    def __init__(
+        self,
+        start,
+        stop,
+        statistic="avg",
+        method="last",
+        *,
+        contact_balance=None,
+        tables="all",
+        direction="both",
+        hours="all",
+        subscriber_subset=None,
+        size=None,
+    ):
+
+        self.start = start
+        self.stop = stop
+        self.direction = direction
+        self.hours = hours
+        self.tables = tables
+        self.method = method
+
+        self.statistic = statistic.lower()
+        if self.statistic not in valid_stats:
+            raise ValueError(
+                "{} is not a valid statistic. Use one of {}".format(
+                    self.statistic, valid_stats
+                )
+            )
+
+        self.distance_matrix_query = DistanceMatrix(level="versioned-cell")
```

Comment:
You mentioned that this was super slow in some scenarios, I'm wondering if this might be why - you don't really need the full n*n matrix of distances here, so it might actually be faster to use `ST_Distance` directly on that reduced set.

I reckon that while in theory the query planner should skip creating the whole matrix, and instead only do it for things which will be included in the join, it probably doesn't because it'll need to do the calculation to work out if it _should_ be in the join.

Reply:
The thinking here was that in this reduced subset calculating `ST_Distance` for each and every pair would be quicker. But in a production scenario the cost of the `DistanceMatrix` would be negligible compared to all distance pairs. The matrix would also probably already be cached because other queries already use it.

Because I wanted to avoid the cost of calculating `ST_Distance` for every pair that I exclude Hartigan from this class. Alternatively, I would consider first getting all distinct pairs, calculating the distances and then joining back to the list of all pairs. And in the case that `DistanceMatrix` is available using it instead for the ML case. Do you think this would be more efficient?

Reply:
I think the latter (get the set of distinct pairs, calculate distances for those, join back) is what I'm suggesting here. There's not a good mechanism to skip doing that iff a distance matrix is applicable and pre-calculated, so my thinking is that not using the distance matrix is the better path, and we can introduce it if this untenable in production.